File: TrabalhoEmGrupo1/TrabalhoFinal/Códigos dos Sorts/C/HeapSort.c
#include <stdio.h>
#include <stdlib.h>

void exchange(int *vet,int i,int j){
  int aux;
  aux = vet[i];
  vet[i] = vet[j];
  vet[j] = aux;
}

void max_heapfy(int *vet,int n,int i){
  int maior=i;
  int esq = 2*i+1;
  int dir = 2*i+2;
  if(esq<n && vet[esq]>vet[i])
    maior = esq;
  if(dir<n && vet[dir]>vet[maior])
    maior = dir;
  if(maior!=i){
    exchange(vet,i,maior);
    max_heapfy(vet,n,maior);
  }
}

void buildMaxHeap(int *vet,int n){
  int i;
  for(i=(n/2);i>=0;i--)
    max_heapfy(vet,n,i);
}

void heapsort(int *vet,int n){
  int i;
  buildMaxHeap(vet,n);
  for(i=n-1;i>=1;i--){
    exchange(vet,0,i);
    max_heapfy(vet,i,0);
  }
}

int main(){
  int *vet,n=0;
  int i;
  int aux;
  FILE *arq;
  arq = fopen("TabelaXX.csv","r");
  if (arq == NULL) {
    printf("Arquivo não encontrado.\n");
    return 0;
  }
  while(fscanf(arq,"%d,",&aux)!= EOF){
    n++;  // Calcula o tamanho do vetor
  }
  vet = malloc(n*sizeof(int));
  rewind(arq); // Retorna ao início do arquivo
  for(i=0;i<n;i++){
    fscanf(arq,"%d",&vet[i]);
  }
  heapsort(vet,n);
  /*for(i=0;i<n;i++)
    printf("%d\n",vet[i]); */ //Imprimindo
  free(vet);
  fclose(arq);
  return 0;
}

File: TrabalhoEmGrupo1/TrabalhoFinal/Códigos dos Sorts/Python/HeapSort.py
import csv

def maxHeapify(vet, n, i):
  # contrói um max heap a partir de vet parcialmente ordenado

  maior = i
  esq = 2*i + 1
  dir = 2*i + 2

  if esq<n and vet[i]<vet[esq]:
    maior = esq

  if dir<n and vet[maior]<vet[dir]:
    maior = dir

  if maior != i:
    vet[i], vet[maior] = vet[maior], vet[i]
    maxHeapify(vet, n, maior)


def buildMaxHeap(vet, n):
  # constrói max heap a partir de vet não ordenado

  i = n//2
  while i >= 0:
    maxHeapify(vet, n, i)
    i -= 1

  
def heapSort():
  # ordena um vetor por heapsort
  
  vet = []

  with open('TabelaXX.csv', 'r') as arq:
    reader = csv.reader(arq)
    for linha in reader:
      vet.append(int(linha[0]))
  # transfere dados para vet
  # converte dados para int
  
  n = len(vet)
  buildMaxHeap(vet, n)
  i = n-1
  while i > 0:
    vet[0], vet[i] = (vet[i], vet[0])
    n -= 1
    maxHeapify(vet, n, 0)
    i -= 1
  

heapSort()